File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # a, b, c, d
        # 1, a, ab, abc
        # 1, c, bc, bcd

        prefix = [1]
        for i in range(len(nums) - 1):
            prefix.append(nums[i] * prefix[i])

        suffix = [1]
        for i in range(len(nums) - 1, 0, -1):
            suffix.append(nums[i] * suffix[len(nums) - i - 1])

        res = []
        for i in range(len(prefix)):
            res.append(prefix[i] * suffix[len(nums) - i - 1])

        return res


        
        

File: Data Structures & Algorithms/duplicate-integer/submission-2.cpp
class Solution {
public:
    bool hasDuplicate(vector<int>& nums) {
        unordered_map<int, bool> m;
        for (int n : nums) {
            if (m.find(n) != m.end()) return true;
            m[n] = true;        
        }
        return false;
    }
};